### PR TITLE
Bohdan/feature/oauth login

### PR DESCRIPTION
**Pull request by BohdanSizomin**

### CODE REVIEW

File: autoapp.py [hunk 1]
```diff
@@ -1,5 +1,11 @@
 # -*- coding: utf-8 -*-
 """Create an application instance."""
 from flaskshop.app import create_app
+from flaskshop.settings import config
+from os import environ as env
 
-app = create_app()
+environment = env.get("FLASK_ENV", "development")
+print(environment)
```
Comment (kipod) [inline autoapp.py]: pls use log(...) inplace print()

File: build.log [hunk 2]
```diff
@@ -0,0 +1,223 @@
+#1 [internal] load build definition from Dockerfile
```
Comment (kipod) [inline build.log:1]: Please put it to .gitignore

File: docker-compose.yml [hunk 3]
```diff
@@ -13,7 +13,7 @@ services:
       PGUSER: ${DB_USER:-postgres}
       PGPASSWORD: ${DB_PASSWD:-secret}
     ports:
-      - 127.0.0.1:${LOCAL_DB_PORT:-5432}:5432
```
Comment (kipod) [inline docker-compose.yml:16]: IMHO you can revert this line

File: docker-compose.yml [hunk 4]
```diff
@@ -39,16 +41,19 @@ services:
 
   web:
     build: .
+    image: simple2b/flaskshop_simple2b
+    volumes:
+      - ${PWD}/.env:/app/.env
```
Comment (kipod) [inline docker-compose.yml]: Hmm... very strange solution... What about using: ```env_file:```  ?

File: docker-compose.yml [hunk 5]
```diff
@@ -39,16 +41,19 @@ services:
...
     depends_on:
       - db
       - redis
       - es
     ports:
       - 127.0.0.1:${LOCAL_WEB_PORT:-5000}:5000
     environment:
-      DB_HOST: db
       REDIS_URI: redis://redis:6379
       ESEARCH_URI: http://es:9200
```
Comment (kipod) [inline docker-compose.yml]: these values need take from .ENV

File: flaskshop/account/models.py [hunk 6]
```diff
@@ -1,28 +1,36 @@
...
+
+
 class User(Model, UserMixin):
+
     __tablename__ = "account_user"
     username = Column(db.String(80), unique=True, nullable=False, comment="user`s name")
     email = Column(db.String(80), unique=True, nullable=False)
     #: The hashed password
-    _password = db.Column(db.String, nullable=False)
+    _password = db.Column(db.String, nullable=True, default=None)
```
Comment (kipod) [inline flaskshop/account/models.py]: No! String in DB must be with a limited length ```db.String(128)```

File: flaskshop/account/models.py [hunk 7]
```diff
@@ -1,28 +1,36 @@
...
     nick_name = Column(db.String(255))
     is_active = Column(db.Boolean(), default=False)
     open_id = Column(db.String(80), index=True)
     session_key = Column(db.String(80), index=True)
+    provider = Column(Enum(OpenidProviders), nullable=False)
 
-    def __init__(self, username, email, password, **kwargs):
-        super().__init__(username=username, email=email, password=password, **kwargs)
+    # def __init__(self, username, email, password, **kwargs):
+    #     super().__init__(username=username, email=email, password=password, **kwargs)
```
Comment (kipod) [inline flaskshop/account/models.py]: please remove these lines

File: flaskshop/account/views.py [hunk 8]
```diff
@@ -10,20 +10,81 @@
...
 import random
 import string
 
 from .forms import AddressForm, LoginForm, RegisterForm, ChangePasswordForm, ResetPasswd
-from .models import UserAddress, User
+from .models import OpenidProviders, UserAddress, User
 from flaskshop.utils import flash_errors
 from flaskshop.order.models import Order
+from flaskshop.logger import log
+import requests
```
Comment (kipod) [inline flaskshop/account/views.py]: wrong import sequence! Please fix!

File: flaskshop/account/views.py [hunk 9]
```diff
@@ -10,20 +10,81 @@
...
+        if not user:
+            user = User.create(
+                open_id=user_data["sub"],
+                email=user_data["email"],
+                is_active=True,
+                username=user_data["email"],
+                provider=OpenidProviders.GOOGLE,
+            )
+        login_user(user)
+        log(log.INFO, f"User logged in.Profile:{user_data}")
```
Comment (kipod) [inline flaskshop/account/views.py]: Don't use f-string in log instruction
```python
log(log.INFO, "User logged in.Profile:[%s]", user_data)
```

File: flaskshop/account/views.py [hunk 10]
```diff
@@ -10,20 +10,81 @@
...
+@csrf_protect.exempt
+def facebook_auth():
+    token = request.json["access_token"]
+    user_id = request.json["user_id"]
+    if token:
+        user_data = requests.get(
+            f"https://graph.facebook.com/{user_id}?fields=id,name,email,picture&access_token={token}"
+        )
+        user_data = user_data.json()
+        user = User.query.filter_by(
```
Comment (kipod) [inline flaskshop/account/views.py:67]: you didn't test if "email" presents in user_data

File: flaskshop/account/views.py [hunk 11]
```diff
@@ -10,20 +10,81 @@
...
+        if not user:
+            user = User.create(
+                open_id=user_data["id"],
+                email=user_data["email"],
+                is_active=True,
+                username=user_data["email"],
+                provider=OpenidProviders.FACEBOOK,
+            )
+        login_user(user)
+        log(log.INFO, f"User logged in.Profile:{user_data}")
```
Comment (kipod) [inline flaskshop/account/views.py]: pls fix

**Review by Nazar-kob — APPROVED** [review]

File: flaskshop/settings.py [hunk 12]
```diff
@@ -7,25 +7,25 @@
...
+#             user=user, passwd=passwd, host=host, db_name=db_name
+#         )
+#     elif db_type == "mysql":
+#         db_uri = "mysql+pymysql://{user}:{passwd}@{host}:{port}/{db_name}?charset=utf8mb4".format(
+#             user=user, passwd=passwd, host=host, port=port, db_name=db_name
+#         )
+#     # redis_uri = "redis://127.0.0.1:6379"
+#     # esearch_uri = "localhost:9200"
+#     redis_uri = "redis://redis:6379"
+#     esearch_uri = "http://localhost:9200"
```
Comment (kipod) [inline flaskshop/settings.py]: pls remove all this commented lines

File: start_server.sh [hunk 13]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+# echo Run db upgrade
+# flask db upgrade
+echo Run app
+flask createdb
+flask seed
+flask run -h 0.0.0.0
```
Comment (kipod) [inline start_server.sh]: the last line must have ```\n```

**Review by kipod — COMMENTED** [review]
Thank you!

File: flaskshop/account/views.py (hunk 10 above)
Comment (BohdanSizomin) [reply flaskshop/account/views.py:67]: In fact, it is impossible not to receive email
It means that if we have already reached that place in the code, then the email will be received 100%

**Pull request merged**